### PR TITLE
tuning research and expansion

### PR DESCRIPTION
**Pull request by BarabanovaIrina**
Добавлено использование библиотеки Hyperopt с возможностью использовать их алгоритм Tree of Parzen Estimators.

### CODE REVIEW

**Comment by nicl-nno** [comment]
А какой-то эффект по качеству настройки TPE дал?

File: core/models/tuning/tuner_adapter.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+
+class HyperoptAdapter(TunerAdapter):
+    def __init__(self, tuner: 'TPE'):
+        super(HyperoptAdapter, self).__init__(tuner)
+        self.model_to_adapt = tuner.trained_model
+        self.data = tuner.tune_data
+        self.params = _search_space_transform(tuner.params_range)
+        self.scorer = tuner.scorer
+        self.best_metric = 0
```
Comment (nicl-nno) [inline core/models/tuning/tuner_adapter.py:26]: А best_metric где-то используется?

File: core/models/tuning/tuners.py [hunk 2]
```diff
@@ -196,3 +199,14 @@ def get_varied_length_range(left_range, right_range):
 
 def _regression_prediction_quality(prediction, real):
     return mse(y_true=real, y_pred=prediction, squared=False)
+
+
+class TPE(Tuner):
```
Comment (nicl-nno) [inline core/models/tuning/tuners.py:204]: TPETuner наверное лучше для единообразия.

File: core/models/tuning/tuner_adapter.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        self.best_params: dict = dict()
+
+    def _objective_function(self, params):
+        self.model_to_adapt.set_params(**params)
+        metric = cross_val_score(self.model_to_adapt,
+                                 self.data.features,
+                                 self.data.target,
+                                 cv=5, scoring=self.scorer).mean()
+
+        if self.data.task.task_type.name == 'classification':
```
Comment (nicl-nno) [inline core/models/tuning/tuner_adapter.py:36]: Тут можно с enum-ом сравнить напрямую.

File: core/models/tuning/tuner_adapter.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            return 1 - metric.mean()
```
Comment (nicl-nno) [inline core/models/tuning/tuner_adapter.py:37]: Это вроде как верно только пока для оценки качества классификации в scorer используется roc auc. 

А нельзя из scorer вытащить значение greater_is_better и опираться на него при проверке?
image

File: core/models/tuning/tuner_adapter.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
+from hyperopt import tpe, fmin, Trials, space_eval
+from sklearn.model_selection import cross_val_score
+import random
+import numpy as np
+
+
+def _search_space_transform(params):
+    space = {}
+    for param_name, param_space in params.items():
+        space[param_name] = hp.choice(param_name, param_space)
```
Comment (nicl-nno) [inline core/models/tuning/tuner_adapter.py:10]: Выпадает "hp not defined"

File: core/models/tuning/tuner_adapter.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+                                 cv=5, scoring=self.scorer).mean()
+
+        if self.data.task.task_type.name == 'classification':
+            return 1 - metric.mean()
+        else:
+            return metric.mean()
+
+    def tune(self, iterations=100):
+        hyperopt_trials = Trials()
+        best_encrypted = fmin(fn=self._objective_function,
```
Comment (nicl-nno) [inline core/models/tuning/tuner_adapter.py:43]: А почему это назвается encrypted?

Comment (BarabanovaIrina) [reply core/models/tuning/tuner_adapter.py:43]: Там такая внутренняя история "dictionary might look like it contains parameters, but it simply contains the encoding used by hyperopt to get the parameter values from the space dictionary." 
И нужно использовать space_eval(), чтобы достать янвые параметры.

File: core/models/tuning/tuner_adapter.py (hunk 1 above)
Comment (BarabanovaIrina) [reply core/models/tuning/tuner_adapter.py:26]: Пока что нигде. Я хотела изначально сохранять лучшую метрику, чтобы потом сравнивать с default, но в hypopt используется исключительно loss. Оставила для доработки.

File: core/models/tuning/tuner_adapter.py (hunk 4 above)
Comment (BarabanovaIrina) [reply core/models/tuning/tuner_adapter.py:37]: Можно, но он находится в protected поле, к которому обращаться будет не очень хорошо. 
Может использовать те классы метрик, которые лежат в core.composer.metrics, добавить им методе scorer, который будет возвращать make_scorer и создать собственный флаг для greater_is_better?

Comment (nicl-nno) [reply core/models/tuning/tuner_adapter.py:37]: Ну эти метрики рассчитаны на работу с цепочками, универсально все равно не сделать.

Мне кажется доступ к _sign тут вполне оправдан, это какой-то косяк дизайна, что cross_val_score не учитывает этот атрибут скорера при выдаче значений.

Можно, конечно, вручную прописывать что-то вроде scorer.greater_is_better = ..., но особого смысла нет.

Алсо, кмк лучше этот словарь назвать tuning_metric_by_type и отправить в tuners.py

**Comment by nicl-nno** [comment]
image


У сейчас почему-то тест test_max_lead_time_in_tune_process падает. Надо убедиться, что он достаточно стабилен.

File: core/models/tuning/tuner_adapter.py (hunk 4 above)
Comment (BarabanovaIrina) [reply core/models/tuning/tuner_adapter.py]: [Перенесено](https://github.com/nccr-itmo/FEDOT/pull/136/commits/e4514a262297b559a0ca6977701ee5ade3fd7273#diff-c73e5c618382254fc5671a4d5e856c59R19-R26) в tuners.py

File: core/models/tuning/tuner_adapter.py (hunk 3 above)
Comment (BarabanovaIrina) [reply core/models/tuning/tuner_adapter.py]: [Исправлено](https://github.com/nccr-itmo/FEDOT/pull/136/commits/e4514a262297b559a0ca6977701ee5ade3fd7273#diff-277a29353b641cf0a3a7dfb2990e01daL36-R37)

File: core/models/tuning/tuners.py (hunk 2 above)
Comment (BarabanovaIrina) [reply core/models/tuning/tuners.py]: [Исправлено](https://github.com/nccr-itmo/FEDOT/pull/136/commits/e4514a262297b559a0ca6977701ee5ade3fd7273#diff-c73e5c618382254fc5671a4d5e856c59L204-R221)

File: core/models/tuning/tuner_adapter.py (hunk 1 above)
Comment (BarabanovaIrina) [reply core/models/tuning/tuner_adapter.py]: Идея была в том, чтобы после выхода из адаптера сравнивать default_score и новый score. И какой лучше, такие параметры и возвращаем (на случай, если каким-то образом найдено не то, что надо)

В результате поле убрано из адаптера. Сравнение с default score производится внутри [тюнера](https://github.com/nccr-itmo/FEDOT/pull/136/commits/e4514a262297b559a0ca6977701ee5ade3fd7273#diff-c73e5c618382254fc5671a4d5e856c59R225-R232). 
Добавленные две новые [функции](https://github.com/nccr-itmo/FEDOT/pull/136/commits/e4514a262297b559a0ca6977701ee5ade3fd7273#diff-c73e5c618382254fc5671a4d5e856c59R57-R67), одна кросс-валидацию оборачивает, вторая сравнивает с default параметрами.

File: core/models/tuning/tuner_adapter.py (hunk 5 above)
Comment (BarabanovaIrina) [reply core/models/tuning/tuner_adapter.py:11]: Исправлено

**Comment by BarabanovaIrina** [comment]
> image
> 
> У сейчас почему-то тест test_max_lead_time_in_tune_process падает. Надо убедиться, что он достаточно стабилен.
Вообще не поняла, что произошло, потому что у меня он стабильно теперь падает даже на CustomTuner. Почему он прошел предыдущие проверки после добавления твоего функционала и после вчерашнего обновления, а теперь падает, мне непонятно и странно. 
Исправила значение assert, но это костыль. Как предложение, убрать вообще этот тест, потому как тест на сам таймер есть, а значит функционал работает и покрыт. А тут по мере разрастания какого-то еще функционала тест так и будет варьироваться.

File: core/models/tuning/tuner_adapter.py (hunk 4 above)
Comment (BarabanovaIrina) [reply core/models/tuning/tuner_adapter.py]: Исправлена проверка [greater_is_better](https://github.com/nccr-itmo/FEDOT/pull/136/commits/64cb630785c8c08433c09088185c44aa07b297ac#diff-277a29353b641cf0a3a7dfb2990e01daL36-R62)

**Comment by nicl-nno** [comment]
Ок, давай этот тест уберём. Странно что он на сервере не падает, конечно.

File: core/models/tuning/tuner_adapter.py (hunk 4 above)
Comment (nicl-nno) [reply core/models/tuning/tuner_adapter.py]: Поменяешь тогда тут же и 1-metric, которую обсуждали?

Comment (BarabanovaIrina) [reply core/models/tuning/tuner_adapter.py]: [Исправлено](https://github.com/nccr-itmo/FEDOT/pull/136/commits/0077679b4acdce220e11ce4e21e9d7a5f5bab537#diff-277a29353b641cf0a3a7dfb2990e01daL37-R37)

**Comment by BarabanovaIrina** [comment]
> Ок, давай этот тест уберём. Странно что он на сервере не падает, конечно.

[Тест](https://github.com/nccr-itmo/FEDOT/pull/136/commits/0077679b4acdce220e11ce4e21e9d7a5f5bab537#diff-4bdd8ef8ff6db41c6faa9d31edfc96d4L152-L172) убран
